### PR TITLE
✨ Add Support for FailureDomains

### PR DESCRIPTION
**Pull request by sf1tzp**
This change adds FailureDomain fields to M3Cluster and M3Machine types, and adds logic to the M3Machine manager code for selecting a host based on a specified FailureDomain.

Related-To: #402
Signed-off-by: Steven Fitzpatrick <steven@fitz.gg>

### CODE REVIEW

**Comment by Rozzii** [comment]
I am very sorry folks, this somehow fallen through the cracks and I've forgot to review it, I am not an expert on the topic so @lentzi90 @kashifest please help us.
/ok-to-test

**Comment by Rozzii** [comment]
@sf1tzp  I hope we are not too late to help you continue this work.

**Review by lentzi90 — COMMENTED** [review]
This looks solid to me!
The `unit` failures look like we would need some additions to the conversion tests because of the API changes. The `generate` failure would probably go away with a rebase and `make generate`.

**Comment by sf1tzp** [comment]
No worries @Rozzii and thanks for the review @lentzi90! I will take a look at this again this evening.

File: baremetal/metal3machine_manager.go [hunk 1]
```diff
@@ -831,6 +833,14 @@ func (m *MachineManager) chooseHost(ctx context.Context) (*bmov1alpha1.BareMetal
...
 		reqs = append(reqs, *r)
 	}
+	if m.Metal3Machine.Spec.FailureDomain != nil {
+		r, err := labels.NewRequirement(FailureDomainLabelPrefix, selection.Equals, []string{*m.Metal3Machine.Spec.FailureDomain})
+		if err != nil {
+			m.Log.Error(err, "Failed to create FailureDomain MatchLabel requirement, not choosing host")
+			return nil, nil, err
+		}
+		reqs = append(reqs, *r)
+	}
```
[the remark this answers is not in the document]
Comment (sf1tzp) [reply baremetal/metal3machine_manager.go:843]: @lentzi90 Sorry, I'm uninitated on conversion-gen. In the `make generate` output, I see

```
hack/tools/bin/conversion-gen \
                --input-dirs=./api/v1alpha5 \
                --output-file-base=zz_generated.conversion  --output-base=/Users/steven/oss/cluster-api-provider-metal3 \
                --go-header-file=./hack/boilerplate/boilerplate.generatego.txt
E0926 16:50:26.490736   38078 conversion.go:756] Warning: could not find nor generate a final Conversion function for github.com/metal3-io/cluster-api-provider-metal3/api/v1beta1.Metal3MachineSpec -> ./api/v1alpha5.Metal3MachineSpec
E0926 16:50:26.491121   38078 conversion.go:757]   the following fields need manual conversion:
E0926 16:50:26.491126   38078 conversion.go:759]       - FailureDomain
```

And in the generated conversion file

``` 
+++ b/api/v1alpha5/zz_generated.conversion.go
@@ -1246,6 +1246,7 @@ func autoConvert_v1beta1_Metal3ClusterStatus_To_v1alpha5_Metal3ClusterStatus(in
        out.LastUpdated = (*v1.Time)(unsafe.Pointer(in.LastUpdated))
        out.FailureReason = (*errors.ClusterStatusError)(unsafe.Pointer(in.FailureReason))
        out.FailureMessage = (*string)(unsafe.Pointer(in.FailureMessage))
+       // WARNING: in.FailureDomains requires manual conversion: does not exist in peer-type
        out.Ready = in.Ready
        // WARNING: in.Conditions requires manual conversion: does not exist in peer-type
        return nil
@@ -1715,14 +1716,10 @@ func autoConvert_v1beta1_Metal3MachineSpec_To_v1alpha5_Metal3MachineSpec(in *v1b
        out.MetaData = (*corev1.SecretReference)(unsafe.Pointer(in.MetaData))
        out.NetworkData = (*corev1.SecretReference)(unsafe.Pointer(in.NetworkData))
        out.AutomatedCleaningMode = (*string)(unsafe.Pointer(in.AutomatedCleaningMode))
+       // WARNING: in.FailureDomain requires manual conversion: does not exist in peer-type
        return nil
 }
```

But I am not sure where to make this manual conversion. I read a bit into it, and found in our v1alpha5 `doc.go`

```
// +k8s:conversion-gen=github.com/metal3-io/cluster-api-provider-metal3/api/v1beta1
```

Which looks right... but do I need to add annotations to the M3Cluster and M3Machine types in v1alpha5 as well?

If you could help point me in the right direction, I would appreciate it! :) Thanks!

[the remark this answers is not in the document]
Comment (lentzi90) [reply baremetal/metal3machine_manager.go:843]: Sure!
Since you are adding a new field here to v1beta1 that does not exist in v1alpha5, conversion-gen does not know what to do with it. We cannot really do anything about it, other than just dropping the field and "taking the responsibility" for it :sweat_smile: . We do that by creating a conversion function and then just calling the automatically generated function, similar to this:

https://github.com/metal3-io/cluster-api-provider-metal3/blob/cde6c510d3f2811b44a8bc528c5420bba3767cca/api/v1alpha5/conversion.go#L219-L232

In other words, you need to add something like

```go
func Convert_v1beta1_Metal3MachineSpec_To_v1alpha5_Metal3MachineSpec(in *v1beta1.Metal3MachineSpec, out *Metal3MachineSpec, s apiconversion.Scope) error {
	// FailureDomain was added in v1beta1 so there is nothing we can do to convert it.
	// It will simply be dropped when converting to v1alpha5.
	return autoConvert_v1beta1_Metal3MachineSpec_To_v1alpha5_Metal3MachineSpec(in, out, s)
}
```

[the remark this answers is not in the document]
Comment (sf1tzp) [reply baremetal/metal3machine_manager.go:843]: Ah okay, thank you! I should have looked at this file, maybe I would have figured it out on my own. One more question though, why are we converting from v1beta1 to v1alpha5? Shouldn't these conversions be forward-looking?

**Comment by sf1tzp** [comment]
Sorry, I'll get these passing locally before I push again ^^

File: baremetal/metal3machine_manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lentzi90) [reply baremetal/metal3machine_manager.go:843]: We actually convert both ways. :slightly_smiling_face:  This is a common pattern, and it works like this: There is one version that the controller handles, that is called the "hub" version. Then there can be any number of other supported versions called "spoken" versions. We add conversion functions to convert between the hub and the spoken versions (not between spoken versions directly, everything is always converted to/from the hub).

This system allows the controller to support any number of API versions simultaneously as long as we can convert them to/from the hub. The user can work with v1alpha5 as long as it is supported even though the controller is actually working with v1beta1. If the user asks for an object using v1alpha5, the webhook automatically converts it before returning it. Similarly, if the user creates an object using v1alpha5, it will be automatically converted and stored as v1beta1.

Kube-builder has some nice documentation around this if you want to read more: https://book.kubebuilder.io/multiversion-tutorial/conversion-concepts

**Comment by lentzi90** [comment]
> @lentzi90 do you know if the unit test in the github action is the same as make unit from the repo? Locally, I keep seeing failures from the Metal3Data reconciler (even on the main branch)

Odd! The unit test is running [hack/unit.sh](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/hack/unit.sh) which should be the same as `make unit-cover-verbose`. Could it be just that it stops before it gets to the fuzzing because of other errors?

File: baremetal/metal3machine_manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sf1tzp) [reply baremetal/metal3machine_manager.go:843]: Thank you for explaining! And for the link. Cheers

**Comment by sf1tzp** [comment]
@lentzi90 sorry to keep you waiting, I want to try this again on my linux box at home. I'll try to get to it this evening!

**Comment by lentzi90** [comment]
> @lentzi90 sorry to keep you waiting, I want to try this again on my linux box at home. I'll try to get to it this evening!

No worries! It should work pretty well also in GitHub Codespaces if you want to try it that way :slightly_smiling_face:

**Comment by Rozzii** [comment]
/remove-lifecycle stale
/lifecycle frozen
IMO we still need this improvement but this issue looks to be stuck at the moment

**Comment by kashifest** [comment]
This PR needs rework. @sf1tzp are you still interested to take this forward? If not, @erjavaskivuori can also take a look

**Comment by Rozzii** [comment]
@kashifest, I have discussed with @sf1tzp on the Issue page https://github.com/metal3-io/cluster-api-provider-metal3/issues/402#issuecomment-3164261736 that we would take over.
I forgot to copy the discussion to this thread.

**Comment by Sunnatillo** [comment]
/close 
Will be taken care of here: https://github.com/metal3-io/cluster-api-provider-metal3/pull/2764
